src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp

```diff
@@ -287,15 +287,15 @@ bool G1BarrierSetC2::expand_barriers(Compile* C, PhaseIterGVN& igvn) const {
 }
 
 uint G1BarrierSetC2::estimated_barrier_size(const Node* node) const {
-  // These Ideal node counts are extracted from the pre-matching Ideal graph
-  // generated when compiling the following method with early barrier expansion:
-  //   static void write(MyObject obj1, Object o) {
-  //     obj1.o1 = o;
-  //   }
   uint8_t barrier_data = MemNode::barrier_data(node);
   uint nodes = 0;
   if ((barrier_data & G1C2BarrierPre) != 0) {
-    nodes += 50;
+    // Only consider the fast path for the barrier that is
+    // actually inlined into the main code stream.
+    // The slow path is laid out separately and does not
+    // directly affect performance.
+    // It has a cost of 4 (Cmp, Bool, If, IfProj).
+    nodes += 4;
```

Comment:
It probably does not make a big overall difference to the loop unrolling heuristics, but for better accuracy you might want to count in the nodes for loading the "active" byte (one node for computing the address relative to the thread-local storage base and one node for the load itself), i.e. 6 nodes in total rather than 4 (the `ThreadLocal` node representing the thread-local storage base is shared by other barrier operations so I would not count it as part of the pre-barrier fast path):

![pre-barrier-fast-path](https://github.com/user-attachments/assets/7718ec94-1612-44ac-b18c-78d962057ab6)


```suggestion
    // It has a cost of 6 (AddP, LoadB, Cmp, Bool, If, IfProj).
    nodes += 6;
```